Task 003

PERECShepard opened the pull request:
Task003 were added

PERECShepard committed:
Update README.md

Task003 were added

PERECShepard committed:
Update game.cpp

cards class were added(not completed)

PERECShepard committed:
Create cards.cpp

cards class were created

PERECShepard committed:
Update CMakeLists.txt

CMake were updated for cards class

ZzOneDay commented on cards.cpp:
```diff
@@ -0,0 +1,69 @@
+public:
+    Cards(Rank rank) : m_rank(rank) { createAndShuffleDesk(desk); }
+
+    std::string getRankToString(const Rank rank) const {
```
This method is better to remove from this. create somethink like cardService

PERECShepard committed:
Update game.cpp

update

PERECShepard committed:
Update cards.cpp

update

ZzOneDay commented:
I want to see:

<---->
Card( 
value: Enum
visible: Boolean
onTable: Boolean
)

<---->
CardPack (
cards: List<Card>
) {
void showCards(){
//show list of card (if visible true [T] if false [#] example)
}

void opencard(int number) {
cards.get(number).visible = true
}

<---->
CardPackService(
randomService
List<Cards> fullPack = initFullPackCards()
) {

fun generateCardPack(int size): CardPack { //size will be 10
cards = new List()
while(card.size < size) {
card = getRandomCard()
}

getRandomCard(): Card {
cardsForTable = fullPack.filter(onTheTable == false).toList()
int randomNumber = randomservice.getRandomNumber(1, cardsForTable.size())
card = cardsForTable.get(randomNumber)
card.setOnTheTable(true)
return card
}

resetFullPack() {
	fullPack.forEach{it.visible = false it.onthetable = false}
}

initFullPackCards(): List<Cards> {
val cards = list()
	I belive you can create this service
	for(Enum enum: yourEnumWithCard) {
	for(i in 1...4) {
	cards.add(card(enum, false, false))
	)
	}
	}
}

ZzOneDay commented on cards.cpp:
```diff
@@ -0,0 +1,82 @@
+        createAndShuffleDesk(desk);
+    }
+
+    std::string getRankToString(const Rank rank) const {
```
Вот смотри у тебя класс карта.  И тут же у тебя в классе метод, который принимает параметр.
Выглядит это так.

У тебя есть друг Вася, и ты просишь его назвать имя другого друга. выглядит как жопка.
(поэтому у тебя должен быть класс карт (ранк, масть если хочешь) и методы (получить масть, получить значение), получить эти значения в строчку. а не так что ты из этого класс делаешь полу сервис, полу хранилище. Дели, либо у тебя это сервис по работе с картами который может переводить из бананов, в огурцы, либо это что у хранилища можно получать данные в разных форманатах.

Либо у тебя должен быть просто класс карты, так стоп. А что это вообще за фигня?

m_rank? это вообще что? Ранк карты? Тогда почему ранк карты у класса "КартЫ". Что это такое?

Почему ты так сопротивляешь делать вирутальный мир на основе реалього? Ты в магазине можешь купить колоды черви? Выглядит как что то очень странное.

ZzOneDay replied:
Если ты все таки хочешь оставить этот метод, и ты его переиспользуешь в других методах, в этом классе которые используются внешне.
То тогда у тебя методы буличные должны быть сверху, и ниже методы прайвет, достыпные только внутри класса этого

ZzOneDay commented on cards.cpp:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    void createAndShuffleDesk(std::vector<std::pair<Rank, bool>>& vec) {
```
Если у тебя метод принимает параметры, то это выглядит как сервис по работе с этими данными. Если ты осущевляешь шафл именно этой колоды, убери паарметры и используй параметры объекта.

Давай определимсся.

У тебя есть классы и объекты. Но ты должен разделять либо это объект типа как хранилище, единица что то бизнессовое типа или сервис по работе с этими объектами, но смешивать не надо

PERECShepard committed:
Update game.cpp

updated: main logic of game and cardService class were added;
few declaration of classes were remove in other classes

PERECShepard committed:
Update player.cpp

player class were updated;
the player now saves the state of his cards

PERECShepard committed:
Update and rename cards.cpp to card.cpp

updated
suit were added;
method to text output were added

PERECShepard committed:
Create cardPack.cpp

upd 1: class were created
upd 2: methods to create full cardPack and 10cardPack were added

PERECShepard committed:
Create cardService.cpp

upd 1: croupier class were added
upd 2: method for main logic of game were added(taking 2 numbers from players and open 2 chosen cards for each player)

PERECShepard committed:
Update CMakeLists.txt

updated for new classes to main game

PERECShepard committed:
Update card.cpp

upd 3: remove redundant members and methods

PERECShepard committed:
Update cardService.cpp

upd 3: calculating scores were added

PERECShepard committed:
Update game.cpp

upd 3: correct scoring were added

ZzOneDay commented:
И так,
Запуск программы не совпадает с заданием.
Expected:
Your balance: 1000
Computer's balance: 1000
You play Choose your card: 
[#][#][#][#][#][#][#][#][#][#]
3 // my input
[#][#][K][#][#][#][#][#][#][#] 
Choose your another card: 
10  // my input
[#][#][K][#][#][#][#][#][#][A] 
Your cards: [K][10] 
Computer play [#][#][#][#][#][#][#][#][#][#] 
Computer chooses: 2 
[#][6][#][#][#][#][#][#][#][#] 
Computer chooses: 8 
[#][6][#][#][#][#][#][#][A][#] 
Computer cards: [6][A]
Random number chosen: 18 
You have [K][A] = 15 
Computer has [6][A] = 17
Computer WIN

next game:
Your balance: 1100
Computer's balance: 900
You play Choose your card: 
[#][#][#][#][#][#][#][#][#][#]
3 

actual:
Now u have: 1000
and pc have: 1000
Please enter number between 1 - 10 range: 2
Please enter number between 1 - 10 range: 2
You entered_ 2 choose another number.
Please enter number between 1 - 10 range: 1
Your cards: 
4♥4♦[#][#][#][#][#][#][#][#]
Pc cards: 
[#][#][#][#]6♣[#][#][#]7♦[#]
Round started
You have: 8
Pc have: 13
Number: 16
PC win because his number was closer
Now u have: 900
and pc have: 1100

ZzOneDay commented:
У компьютера может попасться та же карта что и у меня на столе.
Неправильная работа
Your cards:
9♠K♠[#][#][#][#][#][#][#][#]
Pc cards:
[#][#][#][#][#]9♠[#][#]10♠[#]

ZzOneDay commented:
expected
[#][#][#][#][#][9♠][#][#][10♠][#]
actual
[#][#][#][#][#]9♠[#][#]10♠[#]

ZzOneDay commented on game.cpp:
```diff
@@ -17,36 +19,45 @@ class NumberGame{
+            std::cout << "Your cards: " << std::endl;
+            cardService.showCards(human);
+
+            cardService.getRandomCard(pc);
```
Метод каждый раз инициализрует колоду и не учитывает что какие то карты уже на столе

PERECShepard replied:
я тебя понял
этот нюанс исправлю

ZzOneDay commented on game.cpp:
```diff
@@ -17,36 +19,45 @@ class NumberGame{
 
-            int absComputer = std::abs(sum(pc) - randNumber);
-            int absHuman = std::abs(sum(human) - randNumber);
+            cardService.getRandomCard(human);
```
метод get здесь не подходит. Метод get подразумевает что он вернет что то для работы.

ZzOneDay commented on cardService.cpp:
```diff
@@ -0,0 +1,42 @@
+    CardPack pack;
+public:
+
+    void getRandomCard(Player &player){
```
метод void не может быть get

PERECShepard replied:
исправил на definePlayerChose

ZzOneDay commented on cardPack.cpp:
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include <list>
+#include "card.cpp"
+#include "randomEngine.cpp"
```
Я не мог запустить программу пока не добавил сюда
`#include <iostream>
#include <vector>
#include <algorithm>
#include <random>`

PERECShepard replied:
непонятно почему 
библиотека рандом включена в рандом энжин
алгоритм тут не нужен так как шафл - метод библиотеки рандом 
вектор добавил 
iostream не нужен так как он включен в класс кард а тот в свою очередь в кард пак
это полагаю просто иде у тебя ебуженит (у меня на работе g++ такие финты выкручивает что лишний раз думаешь будто комп сошел с ума)

PERECShepard committed:
Update cardPack.cpp

problem with duplication of card were solved

PERECShepard committed:
Update cardService.cpp

problem with duplication of card were solved

ZzOneDay commented:
Я добавил std::cout << "Size CARDS: " << cardStack.size() << std::endl;
в класс 
std::list<Card> initCardPack() {
        cardPack.clear();

        for (int i = 0; i < 10; ++i) {
            cardPack.push_back(cardStack.top());
            cardStack.pop();
        }

        std::cout << "Size CARDS: " << cardStack.size() << std::endl;
        return cardPack;
    }

Now u have: 1000
and pc have: 1000
Please enter number between 1 - 10 range: 1
Please enter number between 1 - 10 range: 2
Size CARDS: 42
Your cards: 
A♠K♠[#][#][#][#][#][#][#][#]
Size CARDS: 32 // std::cout << "Size CARDS: " << cardStack.size() << std::endl;
Pc cards: 
[#][#]2♠[#][#][#][#]10♥[#][#]
Round started
You have: 27
Pc have: 12
Number: 7
PC win because his number was closer
Now u have: 900
and pc have: 1100
Please enter number between 1 - 10 range: 3
Please enter number between 1 - 10 range: 4
Size CARDS: 230
Your cards:
[#][#]Q♠J♠[#][#][#][#][#][#]
Size CARDS: 220
Pc cards:
[#]3♠[#][#][#][#][#][#]9♥[#]
Round started
You have: 23
Pc have: 12
Number: 13
PC win because his number was closer
Now u have: 800
and pc have: 1200
Please enter number between 1 - 10 range:

С каждым ходом количествао карт в массиве увеличивается. Вероятность встретить теже карты все равно есть. Плюс бесконечнй массив если они будут играть до бесконечно ты грохнешься по памяти)))

PERECShepard committed:
Update cardPack.cpp

duplicating of cards were solved(i hope)

PERECShepard committed:
Update game.cpp

upd

ZzOneDay commented on cardService.cpp:
```diff
@@ -0,0 +1,46 @@
+                std::cout << "[#]";
+            }
+        }
+        std::cout << std::endl;
```
remove this //std::cout << std::endl;

ZzOneDay replied:
Александр, нужно исправлять MR согласно комментариям и резолвить их

PERECShepard committed:
Update game.cpp

output updated

PERECShepard committed:
Update player.cpp

output updated

PERECShepard committed:
Update cardService.cpp

output updated

PERECShepard committed:
Update game.cpp

PERECShepard committed:
type player were added(enum)

PERECShepard committed:
type player were added

PERECShepard committed:
type player were added

ZzOneDay reviewed (approved):
Вроде все гуд

PERECShepard committed:
little fix
